File: tests/accel/accel.c
/*
 * Copyright (c) 2011 Slade Maurer, Alexander Sviridenko
 */

#include <bb/os/kernel/system.h>
#include <bb/os/hardware/drivers/accel/h48c/core.h>

static struct h48c_pins demo_pins = {1, 2, 0, 3};
static bool_t init_complete = 0;
static int is_free_fall;

void
freefall_runner()
{
  static bbos_message_t message; // message to communicate with accelerometer

  // Demo has a new message
  if (bbos_port_receive(FREEFALL, &message) == BBOS_SUCCESS)
    {
      // Try to figure out the command
      switch (message.id)
	{
	case BBOS_DRIVER_OPEN:
	  printf("H48C device has been opened\n");
	  init_complete = 1;
	  break;
	  // Response from the free fall detection
	case H48C_FREEFALL:
	  if (*(int *)message.data == 1)
	    printf("Free fall detected!\n");
	  break;
	default:
	  bbos_panic("Unknown message");
	  break;
	}
    }

  if (!init_complete)
    {
      printf("Send open-message to H48C device driver\n");
      message.id = BBOS_DRIVER_OPEN;
      message.data = &demo_pins;
      bbos_port_send(H48C, &message, DEMO);
      return;
    }

  message.id = H48C_FREE_FALL;
  message.data = &is_free_fall;
  bbos_port_send(H48C, &message, DEMO);
}

int
main()
{
  bbos_init();
  bbos_add_thread(FREEFALL, freefall_runner);
  bbos_add_thread(H48C, h48c_runner);
  bbos_start();

	return 0;
}

File: tests/accel/accel.py
#!/usr/bin/env python

__copyright__ = "Copyright (c) 2011 Sladeware LLC"

import time

from bb import app
from bb.os import get_running_kernel, Kernel, StaticScheduler
from bb.os.hardware.boards import PropellerDemoBoard

kernel = Kernel()
kernel.set_scheduler(StaticScheduler())
h48c = kernel.load_module('bb.os.hardware.drivers.accel.h48c')

device_settings = h48c.H48CDeviceSettings(7, 6, 0, 1)
device = h48c.AccelDevice(settings=device_settings)

def freefall_runner():
    time.sleep(1) # So we humans can see output
    freefall_runner.init_complete = getattr(freefall_runner, 'init_complete', False)
    if not freefall_runner.init_complete:
        if h48c.accel_open(device):
            get_running_kernel().printer("Accelerometer device has been opened")
            freefall_runner.init_complete = True
        return
    if h48c.accel_freefall():
        get_running_kernel().printer("Free fall detected!")
    else:
        get_running_kernel().printer("No free fall")

kernel.add_thread('FREEFALL', freefall_runner)
accel = app.Process("Accel", kernel)
board = PropellerDemoBoard(processes=[accel])

accel.run()

